sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -258,6 +258,24 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
       val outputAttributes = readDataColumns ++ generatedMetadataColumns ++
         partitionColumns ++ constantMetadataColumns
 
+
+      // The metadata attribute references in the filters also have to be categorized as either
+      // constant or generated metadata attributes. Only data filters can contain metadata filters.
+      def categorizeFileSourceMetadataAttributesInFilters(
+          filters: Seq[Expression]): Seq[Expression] =
+        filters.map { filter =>
+          filter.transform {
+            case attr: AttributeReference if FileSourceMetadataAttribute.unapply(attr).isDefined =>
+              if (attr.dataType.asInstanceOf[StructType].fieldNames
+                .forall(fieldName => constantMetadataColumns.exists(fieldName == _.name))) {
```

Comment:
a complex conditional expression,  extracting a function may look clearer

Reply:
I settled for a different implementation